### PR TITLE
Feature/bi 12446/implement free item calculation for admin

### PR DESCRIPTION
**Pull request by daisysanson**
Resolves [BI-12446](https://companieshouse.atlassian.net/browse/BI-12446)

- Adding changes to CostCalculator, MIDController/Service  as well as new helper classes EricAuthoriser and StringHelper to implement logic, allowing users who are logged in with the correct permission (free-mids) to full discount giving them free orders. 
- Adding Unit tests to cover new logic where necessary. 

Screenshot 2024-07-31 at 08 13 31

[BI-12446]: https://companieshouse.atlassian.net/browse/BI-12446?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/service/MissingImageDeliveryCostCalculatorService.java [hunk 1]
```diff
@@ -27,17 +25,28 @@ public class MissingImageDeliveryCostCalculatorService {
...
      * @return all of the relevant costs
      */
-    public ItemCostCalculation calculateCosts(final int quantity, final ProductType productType) {
+    public ItemCostCalculation calculateCosts(final int quantity, final ProductType productType, final boolean userGetsFreeCertificates) {
         checkArguments(quantity);
-        final int calculatedCost = costs.getMissingImageDeliveryItemCost();
-        final String totalItemCost = Integer.toString(quantity * calculatedCost);
+        final int basicCost = costs.getMissingImageDeliveryItemCost();
+        final int calculatedCost;
+        if (userGetsFreeCertificates) {
```
Comment (tombradley1) [inline src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/service/MissingImageDeliveryCostCalculatorService.java:38]: This part of the code could be simplified and made clearer, something like:

`final var totalItemCost = userGetsFreeCertificates ? "0" : Integer.toString(quantity * calculatedCost);`

This way, you no longer need separate variables for discountApplied and itemCost, and you directly handle the final cost based on the user's permission.

File: src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/util/StringHelper.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Component
+public final class StringHelper {
+
+    public Set<String> asSet(String regexDelim, String values) {
+        return Stream.of(Optional.ofNullable(values).orElse("").split(regexDelim))
+                .filter(StringUtils::isNotEmpty)
+                .collect(Collectors.toSet());
```
Comment (artemisgardnerch) [inline src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/util/StringHelper.java:17]: Could this be a static method or class? Also this is just a personal opinion but I've always considered streams to be not the best when it comes to readability, debugging and optimisation.

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/util/StringHelper.java (hunk 2 above)
Comment (daisysanson) [reply src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/util/StringHelper.java:17]: As I'm mocking it in my tests, I won't be able to make use static here so using the Spring @ Component.  
About the streams,  think for the purpose of this small utils method where it's the only thing in the class streams can be handy, cuts down some loads of lines of code?

Comment (artemisgardnerch) [reply src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/util/StringHelper.java:17]: I'm of the opinion that utility methods aren't really what mocking is for, but if you're happy with the design I'm happy to leave it.

Comment (daisysanson) [reply src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/util/StringHelper.java:17]: Fair -  I'll need the component still for autowired part  (so non static) to work but happy to refactor the Eric Authoriser where it is being mocked  [here](https://github.com/companieshouse/missing-image-delivery.orders.api.ch.gov.uk/blob/f93412542b9d77faf68ca2a1b2767abdcf66fe29/src/test/java/uk/gov/companieshouse/missingimagedelivery/orders/api/interceptor/EricAuthoriserTest.java#L28):

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

File: src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/controller/MissingImageDeliveryItemController.java [hunk 3]
```diff
@@ -28,11 +29,8 @@
 import static org.springframework.http.HttpStatus.CREATED;
 import static org.springframework.http.HttpStatus.NOT_FOUND;
 import static org.springframework.http.HttpStatus.OK;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.logging.LoggingUtils.COMPANY_NUMBER_LOG_KEY;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.logging.LoggingUtils.REQUEST_ID_HEADER_NAME;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.logging.LoggingUtils.MISSING_IMAGE_DELIVERY_ID_LOG_KEY;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.logging.LoggingUtils.STATUS_LOG_KEY;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.logging.LoggingUtils.USER_ID_LOG_KEY;
+import static uk.gov.companieshouse.missingimagedelivery.orders.api.logging.LoggingUtils.*;
```
Comment (caismith-ch) [inline src/main/java/uk/gov/companieshouse/missingimagedelivery/orders/api/controller/MissingImageDeliveryItemController.java:32]: Minor comment, likely done via a setting in your IDE but following our styleguides https://github.com/companieshouse/styleguides/blob/main/java.md we should avoid using wildcard imports

File: src/test/java/uk/gov/companieshouse/missingimagedelivery/orders/api/interceptor/EricAuthoriserTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
+package uk.gov.companieshouse.missingimagedelivery.orders.api.interceptor;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
```
Comment (caismith-ch) [inline src/test/java/uk/gov/companieshouse/missingimagedelivery/orders/api/interceptor/EricAuthoriserTest.java:4]: Same as previous comment.

File: src/test/java/uk/gov/companieshouse/missingimagedelivery/orders/api/service/MissingImageDeliveryCostCalculatorServiceTest.java [hunk 5]
```diff
@@ -18,11 +18,7 @@
 import static org.hamcrest.core.Is.is;
 import static org.mockito.Mockito.when;
 import static uk.gov.companieshouse.missingimagedelivery.orders.api.model.ProductType.MISSING_IMAGE_DELIVERY_ACCOUNTS;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.util.TestConstants.CALCULATED_COST;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.util.TestConstants.DISCOUNT_APPLIED;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.util.TestConstants.MISSING_IMAGE_DELIVERY_ITEM_COST;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.util.TestConstants.MISSING_IMAGE_DELIVERY_ITEM_COST_STRING;
-import static uk.gov.companieshouse.missingimagedelivery.orders.api.util.TestConstants.POSTAGE_COST;
+import static uk.gov.companieshouse.missingimagedelivery.orders.api.util.TestConstants.*;
```
Comment (caismith-ch) [inline src/test/java/uk/gov/companieshouse/missingimagedelivery/orders/api/service/MissingImageDeliveryCostCalculatorServiceTest.java:21]: Same as previous comment.

**Review by caismith-ch — APPROVED** [review]

**Comment by ch-code-analysis** [comment]
CI: Security warnings found!

**Pull request merged**